File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int product = nums[0], n = nums.size();
        for(int i = 1, maxx = product, minn = product; i < n; i++){
            if(nums[i] < 0){
                swap(maxx, minn);
            }
            maxx = max(nums[i], maxx * nums[i]);
            minn = min(nums[i], minn * nums[i]);
            product = max(product, maxx);
        }
        return product;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_so_far = nums[0]
        current_max = nums[0]
        current_min = nums[0]
        
        for num in nums[1:]:
            temp = max(num, current_max * num, current_min * num)
            current_min = min(num, current_max * num, current_min * num)
            current_max = temp
            max_so_far = max(max_so_far, current_max)
            
        return max_so_far